core/templates/services/exploration-improvements-task-registrar.service.ts

```diff
@@ -0,0 +1,424 @@
+
+  private initializationChangesAreFlushed: boolean = false;
+
+  constructor(
+      private explorationTaskObjectFactory: ExplorationTaskObjectFactory) {}
+
+  initialize(
+      expId: string,
+      expVersion: number,
+      states: States,
+      expStats: ExplorationStats,
+      openTasks: ExplorationTask[],
+      resolvedTaskTypesByStateName: Map<string, ExplorationTaskType[]>,
+      stateAnswerStats: Map<string, AnswerStats[]>,
+      cstPlaythroughsByStateName: Map<string, CstPlaythrough[]>,
+      eqPlaythroughsByStateName: Map<string, EqPlaythrough[]>,
+      misPlaythroughsByStateName: Map<string, MisPlaythrough[]>) {
+    this.validateInitializeArgs(
+      expId, expVersion, states, expStats, openTasks,
+      resolvedTaskTypesByStateName, stateAnswerStats,
+      cstPlaythroughsByStateName, eqPlaythroughsByStateName,
+      misPlaythroughsByStateName);
+
+    this.expId = expId;
+    this.expVersion = expVersion;
+    this.expStats = expStats;
+
+    const allStateNames = states.getStateNames();
+    this.initializeStatsRegistry(
+      allStateNames, stateAnswerStats, cstPlaythroughsByStateName,
+      eqPlaythroughsByStateName, misPlaythroughsByStateName);
+    this.initializeTaskRegistry(
+      allStateNames, openTasks, resolvedTaskTypesByStateName);
+
+    for (const [stateName, taskRegistry] of this.taskRegistry.entries()) {
+      const statsRegistry = this.statsRegistry.get(stateName);
+      taskRegistry.forEach(t => this.refreshTask(t, statsRegistry));
+    }
+  }
+
+  flushInitializationChanges(): ITaskEntryPayloadDict[] {
+    if (this.initializationChangesAreFlushed) {
+      return [];
+    }
+    this.initializationChangesAreFlushed = true;
+
+    const payload = [];
+    for (const taskRegistry of this.taskRegistry.values()) {
+      if (taskRegistry.didHbrTaskStatusChange()) {
+        payload.push(taskRegistry.hbrTask.toPayloadDict());
+      }
+    }
+    return payload;
+  }
+
+  onStateAdd(stateName: string): void {
+    const newTasks = ImprovementsConstants.TASK_TYPES.map(
+      taskType => this.explorationTaskObjectFactory.createNewObsoleteTask(
+        this.expId, this.expVersion, taskType, stateName));
+    this.taskRegistry.set(stateName, new TaskRegistry(newTasks));
+    this.statsRegistry.set(stateName, new StatsRegistry());
+    newTasks.forEach(t => this.tasksByType.get(t.taskType).push(t));
+  }
+
+  onStateDelete(stateName: string): void {
+    const taskRegistry = this.taskRegistry.get(stateName);
+    taskRegistry.forEach(t => t.markAsObsolete());
+
+    // eslint-disable-next-line dot-notation
+    this.taskRegistry.delete(stateName);
+    // eslint-disable-next-line dot-notation
+    this.statsRegistry.delete(stateName);
+  }
+
+  onStateRename(oldStateName: string, newStateName: string): void {
+    const oldTaskRegistry = this.taskRegistry.get(oldStateName);
+    const newTaskRegistry = (
+      oldTaskRegistry.map(t => t.cloneWithNewTarget(newStateName)));
```

Comment:
I do not think it makes sense to define a whole new "clone()" functionality on the domain object just to handle this use case. It makes it seem like that is a general method that anyone should be able to use. Perhaps you can just do it here using a deepcopy() and changing the relevant field.

Reply:
Hmm I don't think deepcopy would be any better, it sounds way too invasive. Maybe a better name would help? The goal of this method is to discard an obsolete name referenced by a task with a valid replacement, so maybe having the name reflect that would help?

Reply:
Why is deepcopy invasive? All the types of the sub-properties are the same and making a copy of an object is not different from cloning it.

.retarget suggests that you are not copying it, but changing in-place. I also still think you don't need a full-fledged method  for this operation on the object itself.

Reply:
Btw (FYI) angular.copy() is the frontend equivalent of the backend copy.deepcopy(), and we use it quite a bit where this sort of functionality is needed. I'm not sure what the Angular 2+ equivalent is, though.

Reply:
I guess my concern is that it bypasses the constructor, where these tasks check for invariants. Deepcopy would bypass the constructor entirely, allowing us to set properties without having them checked to be valid (note that the properties are readonly, so they can't even be legally changed by the type system).

Constructing a new instance is the only way to create a task with a new target, by design. Given this, I think a method makes most sense, although I do wish there were a notion of "package private" or something like that, so it isn't totally public.

Reply:
I see, got it, thanks. How about a hybrid solution -- drop the clone() method, and assemble the args in the onRename function (or similar), then use the regular constructor to make the new one? Presumably if any errors happen in assembly, the constructor() will detect that?